### DISCUSSION TITLE
How to test retryExceptionPredicate or resultPredicate

### DISCUSSION THREAD

**Discussion by BaruchDavid** [resilience4j/resilience4j#2083] [Q&A]
Hallo, ich habe erfolgreich den Retry implementiert und dabei den resultPredicate bzw. retryExceptionPredicate verwendet.
Mein Ziel dabei ist, dass ich nur bei einem 500 StatusCode den Retry ausführen möchte. 
Nun möchte ich das testen (SpringBoot) und würde gerne verifizieren, dass mein Predicat 3 Mal aufgerufen würde.
Wie kann ich das tun?
Mein Code ist: 

`public class MyAdapter {

      @Retry(name = "feignEx", fallbackMethod = "logFail")
       public String sendPartner(MyRequest request) {
        Response response = myClient.send(request);
        return "ok"
      }
    private String logFail(MyRequest eaziRequest,
                                           Exception exception) {
        return "fail"
    }
}`

and my Predicat is: 

`  @Slf4j
public class  ErrorPredicat implements Predicate<FeignException> {

    @Override
    public boolean test(FeignException exception) {

        if( exception.status() >= HttpStatus.INTERNAL_SERVER_ERROR.value()){      
            return true;
        }
        return false;
    }
}`

and my Config is:

`   
resilience4j:
  retry:
    instances:
      throwingFeignException:
        maxAttempts: 3
        waitDuration: 1s
        exponentialBackoffMultiplier: 2
        resultPredicate: com.my.ErrorPredicat 
        retryExceptionPredicate: com.my.ErrorPredicat 
        failAfterMaxAttempts: true
`

and now my Test, where i would like check that ErrorPredicat.test() was called three times:

`
@SpringBootTest
@ActiveProfiles("test")
class EaziAdapterRetryTest {

    @Autowired
    private RetryRegistry registry;

    @Autowired
    private MyClient myClient;

    @MockBean
    private Sender sender;

    @Test
    void retryTest()  {
       
        Request feignRequest =
            Request.create(Request.HttpMethod.POST, "", Map.of(), "".getBytes(StandardCharsets.UTF_8), null, null);
        FeignException.InternalServerError feignException =
            new FeignException.InternalServerError("", feignRequest, "".getBytes(StandardCharsets.UTF_8), null);
        MyRequest myRequest = new MyRequest();
        when(sender.send(any())).thenThrow(feignException);
        final Retry throwingExceptionRetry = registry.retry("throwingFeignException"); //how can i retrieve my ErrorPredicat here???
        String result = eaziAdapter.sendPartner(myRequest );
        assertThat(result).contains("fail");

    }

}
`

**Comment by RobWin** [comment]
I would usually use Wiremock to test it in a SpringBootTest. 
You can also use Wiremock to simulate two failures and only the third call is successful.
